### PR TITLE
Impl Constants

### PR DESCRIPTION
**Pull request by justkawal**
## Description
- Implements Constants encoding / decoding.

### CODE REVIEW

File: packages/substrate_metadata/lib/models/constant.dart [hunk 1]
```diff
@@ -0,0 +1,8 @@
+part of models;
+
+class Constant {
+  final String type;
```
Comment (Lohann) [inline packages/substrate_metadata/lib/models/constant.dart:4]:
```suggestion
  final Codec type;
```

File: packages/substrate_metadata/lib/models/constant.dart [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+  final Uint8List value;
```
Comment (Lohann) [inline packages/substrate_metadata/lib/models/constant.dart:5]:
```suggestion
  final Uint8List bytes;
```

File: packages/substrate_metadata/lib/models/constant.dart [hunk 3]
```diff
@@ -0,0 +1,8 @@
+part of models;
+
+class Constant {
```
Comment (Lohann) [inline packages/substrate_metadata/lib/models/constant.dart:3]: Please put some docs here

**Review by Lohann — CHANGES_REQUESTED** [review]

File: packages/substrate_metadata/lib/models/constant.dart (hunk 2 above)
Comment (justkawal) [reply packages/substrate_metadata/lib/models/constant.dart:5]: Done

File: packages/polkadart_scale_codec/lib/io/byte_output.dart [hunk 4]
```diff
@@ -31,6 +31,9 @@ class ByteOutput with Output {
   @override
   void write(List<int> bytes) {
     if (_buffer.length - cursor < bytes.length) {
+      // Grow the buffer with the desired size.
+      _buffer.addAll(
+          List.generate((bytes.length - (_buffer.length - cursor)), (_) => 0));
```
[the remark this answers is not in the document]
Comment (justkawal) [reply packages/polkadart_scale_codec/lib/io/byte_output.dart:36]: Done

File: packages/substrate_metadata/lib/models/constant.dart (hunk 3 above)
Comment (justkawal) [reply packages/substrate_metadata/lib/models/constant.dart:5]: Done

File: packages/substrate_metadata/lib/models/constant.dart [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  ///
+  /// The constant type codec suitable for decoding the constant bytes
+  final Codec type;
+
+  ///
+  /// The constant bytes
+  final Uint8List bytes;
+
+  ///
+  /// The constant documentation
```
Comment (Lohann) [inline packages/substrate_metadata/lib/models/constant.dart:15]:
```suggestion
  /// Documentation of the constant.
```

File: packages/substrate_metadata/lib/models/constant.dart [hunk 6]
```diff
@@ -0,0 +1,27 @@
+part of models;
+
+///
+/// A constant object model
```
Comment (Lohann) [inline packages/substrate_metadata/lib/models/constant.dart:4]:
```suggestion
/// Metadata represeting one pallet constant.
```

**Review by Lohann — CHANGES_REQUESTED** [review]

File: packages/polkadart_scale_codec/lib/io/byte_output.dart [hunk 7]
```diff
@@ -31,7 +31,7 @@ class ByteOutput with Output {
   @override
   void write(List<int> bytes) {
     if (_buffer.length - cursor < bytes.length) {
-      _buffer.replaceRange(cursor, cursor + bytes.length, bytes);
+      _buffer.insertAll(cursor, bytes);
```
Comment (Lohann) [inline packages/polkadart_scale_codec/lib/io/byte_output.dart:34]: Good find!!

File: packages/substrate_metadata/lib/models/constant.dart [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+class Constant {
+  ///
+  /// The constant type codec suitable for decoding the constant bytes
+  final Codec type;
```
Comment (Lohann) [inline packages/substrate_metadata/lib/models/constant.dart:8]:
```dart
class Constant<T> {
  ///
  /// The constant type codec suitable for decoding the constant bytes
  final Codec<T> type;
  
  ///
  /// Decoded constant value from the bytes using the type codec
  T get value {
    return type.decode(Input.fromBytes(bytes));
  }
}
```

File: packages/substrate_metadata/lib/models/constant.dart (hunk 6 above)
Comment (justkawal) [reply packages/substrate_metadata/lib/models/constant.dart]: Done

File: packages/substrate_metadata/lib/models/constant.dart (hunk 5 above)
Comment (justkawal) [reply packages/substrate_metadata/lib/models/constant.dart]: Done

File: packages/substrate_metadata/lib/models/constant.dart [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+  ///
+  /// The constant bytes
```
[the remark this answers is not in the document]
Comment (justkawal) [reply packages/substrate_metadata/lib/models/constant.dart]: Done

**Review by Lohann — APPROVED** [review]

**Pull request merged**